src/backend/common/MemoryManager.hpp

```diff
@@ -236,71 +250,75 @@ class MemoryManager
         // Shortcut for empty arrays
         if (!ptr) return;
 
-        lock_guard_t lock(this->memory_mutex);
-        memory_info& current = this->getCurrentMemoryInfo();
+        // Frees the pointer outside the lock.
+        std::unique_ptr<void, std::function<void(void*)>> freed_ptr(nullptr, [this](void* p) {
+                                          this->nativeFree(p);
+                                        });
+        {
+            lock_guard_t lock(this->memory_mutex);
+            memory_info& current = this->getCurrentMemoryInfo();
 
-        locked_iter iter = current.locked_map.find((void *)ptr);
+            locked_iter iter = current.locked_map.find((void *)ptr);
 
-        // Pointer not found in locked map
-        if (iter == current.locked_map.end()) {
-            // Probably came from user, just free it
-            this->nativeFree(ptr);
-            return;
-        }
+            // Pointer not found in locked map
+            if (iter == current.locked_map.end()) {
+                // Probably came from user, just free it
+                freed_ptr.reset(ptr);
```

Comment:
This is a nice trick with `unique_ptr`. :+1:

Reply:
Agreed :+1: